### DISCUSSION TITLE
The time step used by controller [...] is not a multiple of WorldInfo.basicTimeStep!

### DISCUSSION THREAD

**Discussion by omichel** [cyberbotics/webots#5396] [Q&A]
I see this message in the Webots console when running my simulation. What does this mean?

**Comment by omichel** [comment]
Webots allows to specify the control step and the simulation step independently. The control step is the argument passed to the `wb_robot_step function`, it specifies the duration of a step of control of the robot. The simulation step is the value specified in `WorldInfo.basicTimeStep` field, it specifies the duration of a step of integration of the physics simulation, in other words: how often the objects motion must be recomputed. The execution of a simulation step is an atomic operation: it cannot be interrupted. Hence a sensor measurement or a motor actuation must take place between two simulation steps. For that reason the control step specified with each `wb_robot_step` function call must be a multiple of the simulation step. If it is not the case you get this error message. So, for example if the `WorldInfo.basicTimeStep` is 16 (ms), then the control step argument passed to the `wb_robot_step` function can be 16, 32, 48, 64, 80, 128, 1024, etc.

**Reply by cdrizzt** [reply]
Hello,

I have a couple of questions:

Question 1:
What happens if I set 'wb_robot_step' to a value smaller than WorldInfo.basicTimeStep?
For example: WorldInfo.basicTimeStep = 8, wb_robot_step(2)
I tested and found that it behaves the same as when WorldInfo.basicTimeStep = 8, wb_robot_step(8).

Question 2:
Is the execution time of wb_robot_step() related to real-world time?
I noticed that wb_robot_step() is generally consistent with real-world time. However, if I copy a large amount of data after wb_robot_step() and then call wb_robot_step() again, would it cause Webots time to lag behind real time?
